presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayDistinctFunction.java

```diff
@@ -50,42 +56,60 @@ public static Block distinct(@TypeParameter("E") Type type, @SqlType("array(E)")
         }
 
         TypedSet typedSet = new TypedSet(type, array.getPositionCount(), "array_distinct");
-        BlockBuilder distinctElementBlockBuilder = type.createBlockBuilder(null, array.getPositionCount());
+        int distinctCount = 0;
+
+        if (pageBuilder.isFull()) {
+            pageBuilder.reset();
+        }
+
+        BlockBuilder distinctElementBlockBuilder = pageBuilder.getBlockBuilder(0);
         for (int i = 0; i < array.getPositionCount(); i++) {
             if (!typedSet.contains(array, i)) {
                 typedSet.add(array, i);
+                distinctCount++;
                 type.appendTo(array, i, distinctElementBlockBuilder);
             }
         }
 
-        return distinctElementBlockBuilder.build();
+        pageBuilder.declarePositions(distinctCount);
+
+        return distinctElementBlockBuilder.getRegion(distinctElementBlockBuilder.getPositionCount() - distinctCount, distinctCount);
     }
 
     @SqlType("array(bigint)")
-    public static Block bigintDistinct(@SqlType("array(bigint)") Block array)
+    public Block bigintDistinct(@SqlType("array(bigint)") Block array)
     {
         if (array.getPositionCount() == 0) {
             return array;
         }
 
         boolean containsNull = false;
-        LongSet set = new LongOpenHashSet(array.getPositionCount());
-        BlockBuilder distinctElementBlockBuilder = BIGINT.createBlockBuilder(null, array.getPositionCount());
+        TypedSet typedSet = new TypedSet(BIGINT, array.getPositionCount(), "array_distinct");
```

Comment:
With #10794, we don't have to make this specialization on any type with long native container type . Thus we don't have to use `TypedSet` here.

Sorry for the confusion.